neural_compressor/adaptor/torch_utils/teq.py

```diff
@@ -0,0 +1,391 @@
+        Try the model to find the layers which can be smooth quantized.
+        :param op_types: The op types to be smooth quantized
+        :return:
+        """
+        self_absorb_layer = {}
+        for name, module in self.model.named_modules():
+            for op_type in op_types:
+                if op_type == str(module.__class__.__name__):
+                    self_absorb_layer[name] = [name]
+        return self_absorb_layer
+
+    def _get_example_input(self):
+        if self.dataloader == None and self.example_inputs == None:
+            return None
+        if self.example_inputs is None:
+            ##assert self.dataloader, "Please provide dataloader or example_inputs"
+            for idx, x in enumerate(self.dataloader):
+                self.example_inputs = x
+                break
+
+        return self.example_inputs
+
+    @torch.no_grad()
+    def _absorb_scales(self, layer, scale, layer_name=""):
+        """
+        Absorb the scale to the layer at output channel
+        :param layer: The module
+        :param scale: The scale to be absorbed
+        :param layer_name: The layer name
+        """
+        # for insert mul
+        if self.insert_mul:
+            if isinstance(layer, TEQMulLinear):
+                set_module(self.model, layer_name, layer.sq_linear)  ##recover
+            else:
+                new_module = TEQMulLinear(layer, scale)
+                set_module(self.model, layer_name, new_module)
+            return
+
+        if isinstance(layer, torch.nn.BatchNorm2d) or isinstance(layer, torch.nn.GroupNorm) or \
+                isinstance(layer, torch.nn.InstanceNorm2d):
+            if layer.affine:
+                layer.weight *= scale
+                layer.bias *= scale
+            else:
+                layer.affine = True
+                weight = torch.ones(layer.num_features, device=self.device, dtype=self.dtype) * scale
+                layer.weight = torch.nn.Parameter(
+                    weight, requires_grad=False)
+                bias = torch.zeros(layer.num_features, device=self.device, dtype=self.dtype)
+                layer.bias = torch.nn.Parameter(bias, requires_grad=False
+                                                )
+        elif isinstance(layer, torch.nn.LayerNorm):
+            if layer.elementwise_affine:
+                layer.weight *= scale
+                layer.bias *= scale
+            else:
+                layer.elementwise_affine = True
+                weight = torch.ones(layer.num_features, device=self.device, dtype=self.dtype) * scale
+                layer.weight = torch.nn.Parameter(
+                    torch.ones(weight, requires_grad=False))
+                bias = torch.zeros(layer.num_features, device=self.device, dtype=self.dtype)
+                layer.bias = torch.nn.Parameter(
+                    bias, requires_grad=False)
+
+        elif isinstance(layer, torch.nn.Conv2d):
+            ## the order could not be changed
```

Comment:
better consolidate with the code in sq,

Reply:
the code in sq are written in the class, can't merge code. And there are some difference about the code.